<a href="https://colab.research.google.com/github/LianaHuang99/GoogleMLCrashCourse/blob/main/WiDS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np


In [2]:
train_func = pd.read_csv('TRAIN_FUNCTIONAL_CONNECTOME_MATRICES.csv')
train_func.head()

participant_id  0throw_1thcolumn  0throw_2thcolumn  0throw_3thcolumn  \
0   70z8Q2xdTXM3          0.093473          0.146902          0.067893   
1   WHWymJu6zNZi          0.029580          0.179323          0.112933   
2   4PAQp1M6EyAo         -0.051580          0.139734          0.068295   
3   obEacy4Of68I          0.016273          0.204702          0.115980   
4   s7WzzDcmDOhF          0.065771          0.098714          0.097604   

   0throw_4thcolumn  0throw_5thcolumn  0throw_6thcolumn  0throw_7thcolumn  \
0          0.015141          0.070221          0.063997          0.055382   
1          0.038291          0.104899          0.064250          0.008488   
2          0.046991          0.111085          0.026978          0.151377   
3          0.043103          0.056431          0.057615          0.055773   
4          0.112988          0.071139          0.085607          0.019392   

   0throw_8thcolumn  0throw_9thcolumn  ...  195throw_196thcolumn  \
0         -0.035335          0.068583  ...              0.003404   
1          0.077505         -0.004750  ...             -0.008409   
2          0.021198          0.083721  ...              0.053245   
3          0.075030          0.001033  ...             -0.023918   
4         -0.036403         -0.020375  ...              0.066439   

   195throw_197thcolumn  195throw_198thcolumn  195throw_199thcolumn  \
0             -0.010359             -0.050968             -0.014365   
1             -0.008479              0.020891              0.017754   
2             -0.028003              0.028773              0.024556   
3             -0.005356              0.018607              0.016193   
4             -0.076680             -0.047530             -0.031443   

   196throw_197thcolumn  196throw_198thcolumn  196throw_199thcolumn  \
0              0.128066              0.112646             -0.058980   
1              0.094040              0.035141              0.032537   
2              0.166343              0.058925              0.035485   
3              0.072955              0.130135              0.056120   
4              0.221213              0.007343              0.005763   

   197throw_198thcolumn  197throw_199thcolumn  198throw_199thcolumn  
0              0.028228              0.133582              0.143372  
1              0.075007              0.115350              0.138200  
2              0.063661              0.042862              0.162162  
3              0.084784              0.114148              0.190584  
4              0.083820              0.079582              0.067269  

[5 rows x 19901 columns]

In [3]:
train_func.shape

(1213, 19901)

In [4]:
train_func.describe()

0throw_1thcolumn  0throw_2thcolumn  0throw_3thcolumn  0throw_4thcolumn  \
count       1213.000000       1213.000000       1213.000000       1213.000000   
mean           0.060553          0.122315          0.060268          0.041287   
std            0.064178          0.054026          0.057495          0.043491   
min           -0.183279         -0.059932         -0.145566         -0.127827   
25%            0.018482          0.086102          0.026548          0.014457   
50%            0.058276          0.123220          0.061339          0.043246   
75%            0.100103          0.154518          0.099056          0.068408   
max            0.321522          0.390895          0.278429          0.189825   

       0throw_5thcolumn  0throw_6thcolumn  0throw_7thcolumn  0throw_8thcolumn  \
count       1213.000000       1213.000000       1213.000000       1213.000000   
mean           0.069722          0.091007          0.066852          0.000252   
std            0.044222          0.049189          0.046864          0.049046   
min           -0.072043         -0.079184         -0.105722         -0.164297   
25%            0.042462          0.057614          0.036934         -0.031358   
50%            0.067066          0.086494          0.067247          0.002549   
75%            0.096504          0.119404          0.095117          0.031053   
max            0.317500          0.316811          0.270018          0.168196   

       0throw_9thcolumn  0throw_10thcolumn  ...  195throw_196thcolumn  \
count       1213.000000        1213.000000  ...           1213.000000   
mean           0.014128          -0.002914  ...              0.011075   
std            0.038205           0.042462  ...              0.049632   
min           -0.137728          -0.148490  ...             -0.161666   
25%           -0.010635          -0.030538  ...             -0.021376   
50%            0.016130          -0.002604  ...              0.010254   
75%            0.038770           0.024507  ...              0.044165   
max            0.145364           0.128301  ...              0.194616   

       195throw_197thcolumn  195throw_198thcolumn  195throw_199thcolumn  \
count           1213.000000           1213.000000           1213.000000   
mean              -0.004938             -0.004378              0.001610   
std                0.046536              0.042900              0.047424   
min               -0.176523             -0.178688             -0.138048   
25%               -0.033424             -0.033798             -0.030132   
50%               -0.004683             -0.003724              0.000990   
75%                0.024913              0.024007              0.032268   
max                0.183152              0.180562              0.192015   

       196throw_197thcolumn  196throw_198thcolumn  196throw_199thcolumn  \
count           1213.000000           1213.000000           1213.000000   
mean               0.115171              0.049984              0.058144   
std                0.057128              0.051664              0.057674   
min               -0.070094             -0.153540             -0.131455   
25%                0.080291              0.015827              0.022316   
50%                0.113640              0.052705              0.059151   
75%                0.150524              0.082526              0.095192   
max                0.375635              0.228748              0.322084   

       197throw_198thcolumn  197throw_199thcolumn  198throw_199thcolumn  
count           1213.000000           1213.000000           1213.000000  
mean               0.093527              0.089403              0.128946  
std                0.054594              0.058036              0.058490  
min               -0.085566             -0.204160             -0.083077  
25%                0.059621              0.053224              0.090459  
50%                0.093397              0.088612              0.127913  
75%                

In [5]:
train_func.isna().sum()

participant_id          0
0throw_1thcolumn        0
0throw_2thcolumn        0
0throw_3thcolumn        0
0throw_4thcolumn        0
                       ..
196throw_198thcolumn    0
196throw_199thcolumn    0
197throw_198thcolumn    0
197throw_199thcolumn    0
198throw_199thcolumn    0
Length: 19901, dtype: int64

In [6]:
train_quant = pd.read_excel('TRAIN_QUANTITATIVE_METADATA.xlsx')

In [7]:
train_quant.head()

participant_id  EHQ_EHQ_Total  ColorVision_CV_Score  APQ_P_APQ_P_CP  \
0   UmrK0vMLopoR          40.00                    13               3   
1   CPaeQkhcjg7d         -94.47                    14               3   
2   Nb4EetVPm3gs         -46.67                    14               4   
3   p4vPhVu91o4b         -26.68                    10               5   
4   M09PXs7arQ5E           0.00                    14               5   

   APQ_P_APQ_P_ID  APQ_P_APQ_P_INV  APQ_P_APQ_P_OPD  APQ_P_APQ_P_PM  \
0              10               47               13              11   
1              13               34               18              23   
2              10               35               16              10   
3              12               39               19              16   
4              15               40               20              24   

   APQ_P_APQ_P_PP  SDQ_SDQ_Conduct_Problems  SDQ_SDQ_Difficulties_Total  \
0              28                         0                           6   
1              30                         0                          18   
2              29                         1                          14   
3              28                         6                          24   
4              28                         1                          18   

   SDQ_SDQ_Emotional_Problems  SDQ_SDQ_Externalizing  \
0                           1                      5   
1                           6                      8   
2                           2                      8   
3                           4                     16   
4                           4                     11   

   SDQ_SDQ_Generating_Impact  SDQ_SDQ_Hyperactivity  SDQ_SDQ_Internalizing  \
0                          0                      5                      1   
1                          7                      8                     10   
2                          5                      7                      6   
3                          9                     10                      8   
4                          4                     10                      7   

   SDQ_SDQ_Peer_Problems  SDQ_SDQ_Prosocial  MRI_Track_Age_at_Scan  
0                      0                 10                    NaN  
1                      4                  5                    NaN  
2                      4                  9               8.239904  
3                      4                  6                    NaN  
4                      3                  9               8.940679

In [8]:
train_quant.shape

(1213, 19)

In [9]:
train_quant.describe()

EHQ_EHQ_Total  ColorVision_CV_Score  APQ_P_APQ_P_CP  APQ_P_APQ_P_ID  \
count    1213.000000           1213.000000     1213.000000     1213.000000   
mean       58.876917             13.164880        3.781533       13.205276   
std        49.855501              2.780624        1.376700        3.811772   
min      -100.000000              0.000000        0.000000        0.000000   
25%        45.600000             14.000000        3.000000       11.000000   
50%        76.670000             14.000000        3.000000       13.000000   
75%        93.400000             14.000000        4.000000       16.000000   
max       100.000000             14.000000       12.000000       28.000000   

       APQ_P_APQ_P_INV  APQ_P_APQ_P_OPD  APQ_P_APQ_P_PM  APQ_P_APQ_P_PP  \
count      1213.000000      1213.000000     1213.000000     1213.000000   
mean         39.374279        17.785655       16.393240       25.246496   
std           6.245928         3.764112        5.376994        3.950529   
min           0.000000         0.000000        0.000000        0.000000   
25%          36.000000        16.000000       13.000000       23.000000   
50%          40.000000        18.000000       16.000000       26.000000   
75%          43.000000        20.000000       19.000000       28.000000   
max          50.000000        28.000000       37.000000       30.000000   

       SDQ_SDQ_Conduct_Problems  SDQ_SDQ_Difficulties_Total  \
count               1213.000000                 1213.000000   
mean                   2.059357                   12.122836   
std                    2.023322                    6.577207   
min                    0.000000                    0.000000   
25%                    0.000000                    7.000000   
50%                    2.000000                   12.000000   
75%                    3.000000                   17.000000   
max                   10.000000                   34.000000   

       SDQ_SDQ_Emotional_Problems  SDQ_SDQ_Externalizing  \
count                 1213.000000            1213.000000   
mean                     2.307502               7.557296   
std                      2.167549               4.166627   
min                      0.000000               0.000000   
25%                      1.000000               4.000000   
50%                      2.000000               7.000000   
75%                      4.000000              10.000000   
max                     10.000000              20.000000   

       SDQ_SDQ_Generating_Impact  SDQ_SDQ_Hyperactivity  \
count                1213.000000            1213.000000   
mean                    4.073372               5.497939   
std                     2.819731               2.836764   
min                     0.000000               0.000000   
25%                     2.000000               4.000000   
50%                     4.000000               6.000000   
75%                     6.000000               8.000000   
max                    10.000000              10.000000   

       SDQ_SDQ_Internalizing  SDQ_SDQ_Peer_Problems  SDQ_SDQ_Prosocial  \
count            1213.000000            1213.000000        1213.000000   
mean                4.565540               2.258038           7.683430   
std                 3.519634               2.090414           2.190295   
min                 0.000000               0.000000           0.000000   
25%                 2.000000               0.000000           6.000000   
50%                 4.000000               2.000000           8.000000   
75%                 7.000000               4.000000          10.000000   
max                17.000000               9.000000          10.000000   

       MRI_Track_Age_at_Scan  
count             853.000000  
mean               11.245678  
std                 3.234372  
min                 0.000000  
25%                 8.803901  
50%                10.739219  
75%                13.460871  
max                21.564453

There are 1213 patients. MRI_Track_Age_at_Scan has 1213-853=360 NaN

In [10]:
train_categ = pd.read_excel('TRAIN_CATEGORICAL_METADATA.xlsx')

In [11]:
train_categ.head()

participant_id  Basic_Demos_Enroll_Year  Basic_Demos_Study_Site  \
0   UmrK0vMLopoR                     2016                       1   
1   CPaeQkhcjg7d                     2019                       3   
2   Nb4EetVPm3gs                     2016                       1   
3   p4vPhVu91o4b                     2018                       3   
4   M09PXs7arQ5E                     2019                       3   

   PreInt_Demos_Fam_Child_Ethnicity  PreInt_Demos_Fam_Child_Race  \
0                               0.0                            0   
1                               1.0                            2   
2                               1.0                            8   
3                               0.0                            8   
4                               0.0                            1   

   MRI_Track_Scan_Location  Barratt_Barratt_P1_Edu  Barratt_Barratt_P1_Occ  \
0                        1                      21                      45   
1                        3                      15                      15   
2                        1                      18                      40   
3                        3                      15                      30   
4                        3                      15                      20   

   Barratt_Barratt_P2_Edu  Barratt_Barratt_P2_Occ  
0                      21                      45  
1                       0                       0  
2                       0                       0  
3                      18                       0  
4                       0                       0

In [12]:
train_categ.describe()

Basic_Demos_Enroll_Year  Basic_Demos_Study_Site  \
count              1213.000000             1213.000000   
mean               2017.652102                2.014839   
std                   1.122522                1.135147   
min                2015.000000                1.000000   
25%                2017.000000                1.000000   
50%                2018.000000                1.000000   
75%                2019.000000                3.000000   
max                2020.000000                4.000000   

       PreInt_Demos_Fam_Child_Ethnicity  PreInt_Demos_Fam_Child_Race  \
count                       1202.000000                  1213.000000   
mean                           0.424293                     2.080791   
std                            0.687470                     3.164636   
min                            0.000000                     0.000000   
25%                            0.000000                     0.000000   
50%                            0.000000                     0.000000   
75%                            1.000000                     2.000000   
max                            3.000000                    11.000000   

       MRI_Track_Scan_Location  Barratt_Barratt_P1_Edu  \
count              1213.000000             1213.000000   
mean                  2.288541               17.641385   
std                   0.758348                4.004639   
min                   0.000000                0.000000   
25%                   2.000000               15.000000   
50%                   2.000000               18.000000   
75%                   3.000000               21.000000   
max                   4.000000               21.000000   

       Barratt_Barratt_P1_Occ  Barratt_Barratt_P2_Edu  Barratt_Barratt_P2_Occ  
count             1213.000000             1213.000000             1213.000000  
mean                24.892828               14.122012               24.719703  
std                 17.025899                7.201023               17.171113  
min                  0.000000                0.000000                0.000000  
25%                  0.000000               12.000000                5.000000  
50%                 30.000000               18.000000               30.000000  
75%                 40.000000               21.000000               40.000000  
max                 45.000000               21.000000               45.000000

PreInt_Demos_Fam_Child_Ethnicity has 9 NaN

In [13]:
train_categ.shape

(1213, 10)

In [14]:
train_soln = pd.read_excel('TRAINING_SOLUTIONS.xlsx')

In [15]:
train_soln.head()

participant_id  ADHD_Outcome  Sex_F
0   UmrK0vMLopoR             1      1
1   CPaeQkhcjg7d             1      0
2   Nb4EetVPm3gs             1      0
3   p4vPhVu91o4b             1      1
4   M09PXs7arQ5E             1      1

In [16]:
train_soln.shape

(1213, 3)

In [18]:
train_soln['ADHD_Outcome'].sum()

831

In [19]:
train_soln['Sex_F'].sum()

416

In [28]:
F_ADHD = train_soln[(train_soln['ADHD_Outcome']==1) & (train_soln['Sex_F']==1)].shape[0]/831*100
print('Female ADHD%:', F_ADHD)
print('Male ADHD%:', 100-F_ADHD)

Female ADHD%: 30.084235860409148
Male ADHD%: 69.91576413959085


## Merge fMRI and Solutions

In [30]:
train_func_soln = train_func.merge(train_soln, on='participant_id')

In [34]:
test_func = pd.read_csv('TEST_FUNCTIONAL_CONNECTOME_MATRICES.csv')
test_categ = pd.read_excel('TEST_CATEGORICAL.xlsx')
test_quant = pd.read_excel('TEST_QUANTITATIVE_METADATA.xlsx')

In [32]:
features =list(train_func_soln.columns)
target=['ADHD_Outcome','Sex_F']
features.remove('ADHD_Outcome')
features.remove('Sex_F')
features

['participant_id',
 '0throw_1thcolumn',
 '0throw_2thcolumn',
 '0throw_3thcolumn',
 '0throw_4thcolumn',
 '0throw_5thcolumn',
 '0throw_6thcolumn',
 '0throw_7thcolumn',
 '0throw_8thcolumn',
 '0throw_9thcolumn',
 '0throw_10thcolumn',
 '0throw_11thcolumn',
 '0throw_12thcolumn',
 '0throw_13thcolumn',
 '0throw_14thcolumn',
 '0throw_15thcolumn',
 '0throw_16thcolumn',
 '0throw_17thcolumn',
 '0throw_18thcolumn',
 '0throw_19thcolumn',
 '0throw_20thcolumn',
 '0throw_21thcolumn',
 '0throw_22thcolumn',
 '0throw_23thcolumn',
 '0throw_24thcolumn',
 '0throw_25thcolumn',
 '0throw_26thcolumn',
 '0throw_27thcolumn',
 '0throw_28thcolumn',
 '0throw_29thcolumn',
 '0throw_30thcolumn',
 '0throw_31thcolumn',
 '0throw_32thcolumn',
 '0throw_33thcolumn',
 '0throw_34thcolumn',
 '0throw_35thcolumn',
 '0throw_36thcolumn',
 '0throw_37thcolumn',
 '0throw_38thcolumn',
 '0throw_39thcolumn',
 '0throw_40thcolumn',
 '0throw_41thcolumn',
 '0throw_42thcolumn',
 '0throw_43thcolumn',
 '0throw_44thcolumn',
 '0throw_45thcolumn',


In [39]:
features.remove('participant_id')


In [40]:
X = train_func_soln[features]
y = train_func_soln[target]

In [41]:
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [42]:
from sklearn.multioutput import MultiOutputClassifier
from sklearn.ensemble import RandomForestClassifier

# Create the base model (Random Forest)
base_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Multi-output classification wrapper
multi_target_model = MultiOutputClassifier(base_model)

# Train the model
multi_target_model.fit(X_train, y_train)

# Predict on train and validation data
y_train_pred = multi_target_model.predict(X_train)
y_valid_pred = multi_target_model.predict(X_valid)


In [46]:
from sklearn.metrics import accuracy_score, classification_report, f1_score

print("Classification Report for ADHD:")
print(classification_report(y_train['ADHD_Outcome'], y_train_pred[:, 0]))

print("Classification Report for Sex:")
print(classification_report(y_train['Sex_F'], y_train_pred[:, 1]))

# Generate combined report for both outputs
print("Overall Classification Report:")
print(classification_report(y_train, y_train_pred, target_names=['ADHD_Outcome', 'Sex_F']))

Classification Report for ADHD:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       310
           1       1.00      1.00      1.00       660

    accuracy                           1.00       970
   macro avg       1.00      1.00      1.00       970
weighted avg       1.00      1.00      1.00       970

Classification Report for Sex:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       644
           1       1.00      1.00      1.00       326

    accuracy                           1.00       970
   macro avg       1.00      1.00      1.00       970
weighted avg       1.00      1.00      1.00       970

Overall Classification Report:
              precision    recall  f1-score   support

ADHD_Outcome       1.00      1.00      1.00       660
       Sex_F       1.00      1.00      1.00       326

   micro avg       1.00      1.00      1.00       986
   macro avg       1.00      1.00 

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in samples with no true nor predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [45]:
# Generate classification report for ADHD and Sex separately
print("Classification Report for ADHD:")
print(classification_report(y_valid['ADHD_Outcome'], y_valid_pred[:, 0]))

print("Classification Report for Sex:")
print(classification_report(y_valid['Sex_F'], y_valid_pred[:, 1]))

# Generate combined report for both outputs
print("Overall Classification Report:")
print(classification_report(y_valid, y_valid_pred, target_names=['ADHD_Outcome', 'Sex_F']))

Classification Report for ADHD:
              precision    recall  f1-score   support

           0       0.47      0.10      0.16        72
           1       0.71      0.95      0.82       171

    accuracy                           0.70       243
   macro avg       0.59      0.53      0.49       243
weighted avg       0.64      0.70      0.62       243

Classification Report for Sex:
              precision    recall  f1-score   support

           0       0.63      0.99      0.77       153
           1       0.00      0.00      0.00        90

    accuracy                           0.63       243
   macro avg       0.31      0.50      0.38       243
weighted avg       0.40      0.63      0.48       243

Overall Classification Report:
              precision    recall  f1-score   support

ADHD_Outcome       0.71      0.95      0.82       171
       Sex_F       0.00      0.00      0.00        90

   micro avg       0.71      0.62      0.67       261
   macro avg       0.36      0.48 

/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in samples with no true nor predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [47]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(multi_target_model, X, y, cv=5, scoring='f1_macro')
print("Cross-Validation F1 Scores:", scores)
print("Mean F1 Score:", scores.mean())


Cross-Validation F1 Scores: [0.42293233 0.41504854 0.43407895 0.41839746 0.42171717]
Mean F1 Score: 0.4224348905840312
